Article > Api, Application : 아티클 단건 조회(단건/목록), 좋아요 등록 및 삭제 기능 추가

wowddok99 opened the pull request:
## Issues

- Resolves #172 
- Resolves #173
- Resolves #174 
- Resolves #175 
- Resolves #176 
- Resolves #181

## Description

- #97 
  - 아티클 조회(단건/목록) 및 좋아요 등록·삭제 기능에 대한 API 및 Application 계층 구현

<br />

## Review Points

- 전체적인 흐름이나 구조 위주로 봐주시면 감사하겠습니다.

<br />

## How Has This Been Tested?

- 테스트 생략

wowddok99 committed:
feat(article): add api and application layers for like and read features

- Added api(domain, readmodel) and application(usecase, port) layers for:
  - article like register/delete
  - article single and list read
- Changed Article domain ID type from Long to String

wch-os reviewed (commented):
Application 모듈에서의 port와 usecase 기능 정의는 잘 해주신 것 같은데.. 혹시 service 로직은 미구현 상태인걸까요??

wowddok99 commented:
> Application 모듈에서의 port와 usecase 기능 정의는 잘 해주신 것 같은데.. 혹시 service 로직은 미구현 상태인걸까요??

앗 확인해보니 누락이네요..! 꼼꼼하게 리뷰해주셔서 감사합니다.
추가로 작성해서 올리겠습니다! 감사합니다.

wowddok99 committed:
feat(article): add ArticleQueryService for fetching article detail and list by blogId

wowddok99 committed:
feat(article-like): implement ArticleLikesCommandService for like create and delete

- Added ArticleLikesCommandService to handle like creation and deletion logic
- Changed save method in ArticleLikesCommandRepositoryPort to return ArticleLike
- Cleaned up formatting in port and usecase interfaces

wowddok99 commented:
> Application 모듈에서의 port와 usecase 기능 정의는 잘 해주신 것 같은데.. 혹시 service 로직은 미구현 상태인걸까요??

말씀해주신 부분 수정하여 커밋했습니다!
감사합니다😀

merge-simpson commented on services/article/api/domain/src/main/java/nettee/article/domain/ArticleLike.java:
```diff
@@ -0,0 +1,31 @@
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class ArticleLike {
```
이하 애노테이션도 고려해 보세요.

```diff
+ @ToString
+ @EqualsAndHashCode
```

merge-simpson commented on services/article/api/domain/src/main/java/nettee/article/domain/ArticleLike.java:
```diff
@@ -0,0 +1,31 @@
+    private Instant createdAt;
+
+    private Instant updatedAt;
+}
```
일부 속성 수정을 위한 메서드들이 필요해 보여요.

예를 들어, 지난 번에 논의한 '좋아요 카운트를 N개 추가'하는 메서드는, 애플리케이션 계층에선 유효할 거예요.

```java
public void increaseCount(int amount) {
    // (중략)
    this.count += amount;
}
```

merge-simpson commented on services/article/api/domain/src/main/java/nettee/article/domain/ArticleLikeStatus.java:
```diff
@@ -0,0 +1,6 @@
+
+public enum ArticleLikeStatus {
+    ACTIVE,
+    DELETED
```
**💊 매핑용 테이블이므로 하드 딜리트 가능해 보입니다.**

팀 내에서 작업 방식, 데이터 보존 수위를 논의해야 정할 수 있어 보입니다.
그리고 만약 소프트 딜리트를 채택한다면 `DELETED` 대신 `REMOVED`를 사용하고 있습니다.

```suggestion
public enum ArticleLikeStatus {
    /**
     * 정상 (활성화)
     */
    ACTIVE,
    /**
     * 어뷰징 감지 등으로 이 사용자의 '좋아요' 카운트를 일시적 또는 영구적으로 비활성화
     */
    SUSPENDED
```

merge-simpson replied:
`REMOVED`까지는 관례적으로 추가할 수 있어 보입니다.

merge-simpson commented on services/article/application/src/main/java/nettee/article/port/ArticleLikesCommandRepositoryPort.java:
```diff
@@ -0,0 +1,10 @@
+
+    ArticleLike save(String userId, String profileId, String articleId);
+
+    void delete(String userId, String profileId, String articleId);
```
`userId`는 `profileId`에 있어 유일하기 때문에, `profileId`만 있어도 충분할 것 같습니다.

merge-simpson replied:
@wowddok99 님, 정정합니다...!

여긴 usecase가 아니라 port여서 `userId`도 받아야 하네요.

merge-simpson commented on services/article/application/src/main/java/nettee/article/service/ArticleLikesCommandService.java:
```diff
@@ -0,0 +1,27 @@
+
+@Service
+@RequiredArgsConstructor
+@Transactional
```
트랜잭션을 서비스 전체에 감싸지 않는 것이 좋아 보입니다.

```suggestion
```

merge-simpson commented on services/article/application/src/main/java/nettee/article/usecase/ArticleLikeDeleteUseCase.java:
```diff
@@ -0,0 +1,6 @@
+
+public interface ArticleLikeDeleteUseCase {
+
+    void deleteArticleLike(String userId, String profileId, String articleId);
```
`userId` 대신 `profileId`만 있어도 좋을 것 같습니다.

> 아직은 1 계정 1 프로필이므로, 프로필 단위로 삭제하는 기능만 있어도 좋아 보입니다.
> 나중에 use case가 추가되면 `deleteByUserIdAndArticleId(String userId, String articleId)` 메서드를 추가할 수 있어 보여요.

```suggestion
    void deleteByProfileIdAndArticleId(String profileId, String articleId);
```

wowddok99 replied:
@merge-simpson 님! 정성스러운 리뷰 감사드립니다 : )
방금 말씀해주신 부분 반영하여 커밋했습니다!
감사합니다!

merge-simpson reviewed (commented):
빠른 작업 감사드립니다!

아직 조금 더 수정할 부분이 있어서 리뷰 남겨 드립니다!
수정 후 다시 리뷰하면 좋을 것 같습니다.

wowddok99 committed:
refactor(article): Apply review feedback for Article Likes feature

merge-simpson commented on services/article/api/domain/src/main/java/nettee/article/domain/ArticleLike.java:
```diff
@@ -0,0 +1,44 @@
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class ArticleLike {
```
### 💊 클래스 이름을 `ArticleLikes`로 하는 건 어떻게 생각하시나요?

흔히 사용하는 RDBMS에서 `LIKE`는 키워드로 사용되므로 `likes` 같은 테이블·컬럼 이름을 주로 사용합니다.  
(물론 `"like"`를 사용할 수도 있습니다.)

> 개인적으로 `likes`는 이미 한 단어처럼 사용된다고 생각합니다.
> 마치 `views`처럼요.

merge-simpson commented on services/article/api/domain/src/main/java/nettee/article/domain/ArticleLike.java:
```diff
@@ -0,0 +1,44 @@
+            throw new IllegalArgumentException("Amount must be positive.");
+        }
+
+        this.count += amount;
+    }
```
💊 상태를 바꾸기 위한 메서드도 필요해 보입니다.

방식은 다양하고 따로 제약을 두진 않습니다.

### 1. 상태별 메서드 제어

```suggestion
        this.count += amount;
    }
    
    public void activate() {
        this.status = ArticleLikesStatus.ACTIVE;
    }
    
    public void suspend() {
        this.status = ArticleLikesStatus.SUSPENDED;
    }
    
    public void softDelete() {
        this.status = ArticleLikesStatus.REMOVED;
    }
```

### 2. 상태 업데이트 메서드

```suggestion
        this.count += amount;
    }
    
    public void updateStatus(ArticleLikesStatus status) {
        Objects.requireNonNull(status, "status must not be null.");
        this.status = status;
    }
```

merge-simpson commented on services/article/application/src/main/java/nettee/article/port/ArticleLikesCommandRepositoryPort.java:
```diff
@@ -0,0 +1,10 @@
+import nettee.article.domain.ArticleLike;
+
+public interface ArticleLikesCommandRepositoryPort {
+
+    ArticleLike save(String profileId, String articleId);
+
```
제가 [이전 리뷰](https://github.com/nettee-space/nettee-blolet-backend/pull/182#discussion_r2263672948)에서 실수한 부분이 있습니다.

정정한 내용은 다음과 같습니다:

- 서비스는 `userId`를 입력받지 않아도 되지만 (`profileId`를 바탕으로 auth 서버에서 `userId`를 받아 오는 게 안전하기 때문)
- Port는 `userId`를 입력받아야 합니다.

또는 포트가 `ArticleLikes` 객체를 입력받는 방식이 되면 어떨까 합니다.

merge-simpson commented on services/article/application/src/main/java/nettee/article/service/ArticleLikesCommandService.java:
```diff
@@ -0,0 +1,26 @@
+@RequiredArgsConstructor
+public class ArticleLikesCommandService implements ArticleLikeCreateUseCase, ArticleLikeDeleteUseCase {
+
+    private final ArticleLikesCommandRepositoryPort articleLikesCommandRepository;
+
+    @Override
+    public ArticleLike createArticleLike(String profileId, String articleId) {
+        return articleLikesCommandRepository.save(profileId, articleId);
+    }
```
제가 [이전 리뷰](https://github.com/nettee-space/nettee-blolet-backend/pull/182#discussion_r2263672948)에서 실수한 부분이 있습니다.

정정한 내용은 다음과 같습니다:

- 서비스는 `userId`를 입력받지 않아도 되지만 (`profileId`를 바탕으로 auth 서버에서 `userId`를 받아 오는 게 안전하기 때문)
- Port는 `userId`를 입력받아야 합니다.

또는 포트가 `ArticleLikes` 객체를 입력받는 방식이 되면 어떨까 합니다.

merge-simpson reviewed (approved):
👍 빠른 작업 감사합니다.

이전 리뷰에서 한 부분 착오가 있었습니다.
다시 이야기된 부분들을 수정한 다음, 우선 병합해 주시면 좋을 것 같습니다!

merge-simpson commented on services/article/api/domain/src/main/java/nettee/article/domain/ArticleLikeStatus.java:
```diff
@@ -0,0 +1,7 @@
+package nettee.article.domain;
+
+public enum ArticleLikeStatus {
```
여기서도 `ArticleLikesStatus`로 명명하시는 거 어떻게 생각하시나요?
(`Likes`로 `s` 추가)

wowddok99 committed:
fix: align articleLikes naming with project specifications
